thumbnails/thumbnails.go

```diff
@@ -2,114 +2,150 @@ package thumbnails
 	builder := &bytes.Buffer{}
 	_, err = builder.WriteString("WEBVTT\n")
 	if err != nil {
 		return err
 	}
-	timestamp := time.Time{}
-	outputLocation := output.JoinPath("thumbnails").String()
-	for i, file := range files {
-		// we skip the first entry because we need to calculate the end timestamp of each frame by looking
-		// at the PTS of the next thumbnail, we then refer to the previous file when uploading at the bottom of the loop.
-		// e.g. keyframe_0.jpg we know starts at 00:00:00 but we need to check the PTS of the next file to know the end time.
-		if i == 0 {
-			continue
-		}
-		// extract the PTS value from the filename to be able to calculate the timestamp of the frame
-		filename := path.Base(file)                                        // e.g. keyframes_0.jpg
-		withoutExt := strings.TrimSuffix(filename, filepath.Ext(filename)) // e.g. keyframes_0
-		parts := strings.Split(withoutExt, "_")
-		if len(parts) < 2 {
-			return fmt.Errorf("thumbnail filename should contain an underscore %s", filename)
-		}
-		pts, err := strconv.ParseFloat(parts[1], 64)
-		if err != nil {
-			return fmt.Errorf("couldn't parse float for thumbnails %w", err)
-		}
-		seconds := pts / videoTrack.FPS
-
-		const layout = "15:04:05.000"
-		start := timestamp.Format(layout)
-		timestamp = time.Time{}.Add(time.Duration(seconds) * time.Second)
-		end := timestamp.Format(layout)
-
-		previousFile := files[i-1]
-		_, err = builder.WriteString(fmt.Sprintf("%s --> %s\n%s\n\n", start, end, path.Base(previousFile)))
+	var currentTime time.Time
+	// loop through each segment, generate a thumbnail image and upload it to storage
+	for _, segment := range mediaPlaylist.GetAllSegments() {
+		thumbOut, err := processSegment(requestID, inputURL, segment, tempDir, outputLocation)
 		if err != nil {
 			return err
 		}
 
-		fileReader, err := os.Open(previousFile)
+		start := currentTime.Format(layout)
+		currentTime = currentTime.Add(time.Duration(segment.Duration) * time.Second)
+		end := currentTime.Format(layout)
+		_, err = builder.WriteString(fmt.Sprintf("%s --> %s\n%s\n\n", start, end, path.Base(thumbOut)))
 		if err != nil {
 			return err
 		}
-		defer fileReader.Close()
-		err = clients.UploadToOSURL(outputLocation, path.Base(previousFile), fileReader, time.Minute)
-		if err != nil {
-			return fmt.Errorf("failed to upload thumbnail: %w", err)
-		}
 	}
 
 	err = clients.UploadToOSURLFields(outputLocation, "thumbnails.vtt", builder, time.Minute, &drivers.FileProperties{ContentType: "text/vtt"})
 	if err != nil {
 		return fmt.Errorf("failed to upload vtt: %w", err)
 	}
+	log.Log(requestID, "generate thumbs end", "input", input)
 	return nil
 }
+
+func processSegment(requestID string, inputURL *url.URL, segment *m3u8.MediaSegment, tempDir string, outputLocation string) (string, error) {
+	segURL := inputURL.JoinPath("..", segment.URI)
+	var (
+		rc  io.ReadCloser
+		err error
+	)
+	// download segment
+	err = backoff.Retry(func() error {
```

Comment:
We could save them to disk at the point of segmenting yes but that wouldn't apply to recordings where it's already segmented.
Just thought though I had tried in previous iterations of the catalyst-uploader to pass the URL to ffmpeg instead, I'll try that again here (I had dismissed it because I saw some issues previously but never dug into it)